### PR TITLE
[SPARK-38003][SQL] LookupFunctions rule should only look up functions from the scalar function registry

### PR DESCRIPTION
**Pull request by allisonwang-db**
### What changes were proposed in this pull request?
This PR updates the Analyzer rule `LookupFunctions` to only use scalar function registry instead of using both scalar function and table function registries, since currently LookupFunctions only handles scalar functions.

### Why are the changes needed?

To make the error message consistent when a scalar function does not exist.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Existing unit tests.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -2114,6 +2126,14 @@ class Analyzer(override val catalogManager: CatalogManager)
       }
     }
 
+    def lookupBuiltinOrTempTableFunction(name: Seq[String]): Option[ExpressionInfo] = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:2129]: We also need to call this method when resolving `UnresolvedFunc`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 2]
```diff
@@ -2006,21 +2006,33 @@ class Analyzer(override val catalogManager: CatalogManager)
...
+          nameParts: Seq[String],
+          lookupBuiltinOrTempFunc: Seq[String] => Option[ExpressionInfo]): T = {
+        if (lookupBuiltinOrTempFunc(nameParts).isDefined) {
+          f
+        } else {
+          val CatalogAndIdentifier(catalog, ident) = expandIdentifier(nameParts)
+          val fullName = normalizeFuncName(catalog.name +: ident.namespace :+ ident.name)
+          if (externalFunctionNameSet.contains(fullName)) {
+            f
+          } else if (catalog.asFunctionCatalog.functionExists(ident)) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:2020]: this checks both scalar and table functions, shall we differentiate them?

One problem is: only the v1 catalog supports table function, so there is no v2 API for it. The logic will be a bit complicatd
```
if (isSessionCatalog) {
  sessionCatalog.tableFunctionExist(...)
} else {
  false
}
```

Comment (allisonwang-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Let me update this PR to only fix the bug in looking up scalar functions. Also, I discovered that the current error message for table functions is more user-friendly: `could not resolve `dummy` to a table-valued function`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala [hunk 3]
```diff
@@ -1553,18 +1553,23 @@ class SessionCatalog(
...
+        lookupTempFuncWithViewContext(name, isBuiltinFunction, functionRegistry.lookupFunction))
     }
   }
 
+  /**
+   * Look up the `ExpressionInfo` of the given table function by name if it's a
+   * built-in or temp function.
+   */
+  def lookupBuiltinOrTempTableFunction(name: String): Option[ExpressionInfo] = synchronized {
+    lookupTempFuncWithViewContext(name, isBuiltinFunction, tableFunctionRegistry.lookupFunction)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:1570]: shall we pass the correct `isBuiltinFunction` function to only check table functions?

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**